perun-oidc-server/src/main/java/cz/muni/ics/oauth2/web/DeviceEndpoint.java

```diff
@@ -164,138 +210,230 @@ public String requestDeviceCode(@RequestParam("client_id") String clientId, @Req
 
-		session.setAttribute("authorizationRequest", authorizationRequest);
-		session.setAttribute("deviceCode", dc);
+		session.setAttribute(AUTHORIZATION_REQUEST, authorizationRequest);
+		session.setAttribute(DEVICE_CODE_SESSION_ATTRIBUTE, dc);
 
-		return "approveDevice";
+		return tryToSetThemedApproveDevice(model, p, req, (DeviceCode) model.get(DC));
 	}
 
 	@PreAuthorize("hasRole('ROLE_USER')")
 	@RequestMapping(value = "/" + USER_URL + "/approve", method = RequestMethod.POST)
-	public String approveDevice(@RequestParam("user_code") String userCode,
-								@RequestParam(value = "user_oauth_approval") Boolean approve,
-								ModelMap model, Authentication auth, HttpSession session) {
-		AuthorizationRequest authorizationRequest = (AuthorizationRequest) session.getAttribute("authorizationRequest");
-		DeviceCode dc = (DeviceCode) session.getAttribute("deviceCode");
+	public String approveDevice(@RequestParam(USER_CODE) String userCode,
+								@RequestParam(value = USER_OAUTH_APPROVAL) Boolean approve,
+								Principal p,
+								HttpServletRequest req,
+								ModelMap model,
+								Authentication auth,
+								HttpSession session) {
+
+		AuthorizationRequest authorizationRequest = (AuthorizationRequest) session.getAttribute(AUTHORIZATION_REQUEST);
+		DeviceCode dc = (DeviceCode) session.getAttribute(DEVICE_CODE_SESSION_ATTRIBUTE);
 
 		// make sure the form that was submitted is the one that we were expecting
 		if (!dc.getUserCode().equals(userCode)) {
-			model.addAttribute("error", "userCodeMismatch");
-			return "requestUserCode";
+			model.addAttribute(ERROR, USER_CODE_MISMATCH);
+			return REQUEST_USER_CODE;
 		}
 
 		// make sure the code hasn't expired yet
 		if (dc.getExpiration() != null && dc.getExpiration().before(new Date())) {
-			model.addAttribute("error", "expiredUserCode");
-			return "requestUserCode";
+			model.addAttribute(ERROR, EXPIRED_USER_CODE);
+			return REQUEST_USER_CODE;
 		}
 
 		ClientDetailsEntity client = clientService.loadClientByClientId(dc.getClientId());
-
-		model.put("client", client);
+		model.put(CLIENT, client);
 
 		// user did not approve
 		if (!approve) {
-			model.addAttribute("approved", false);
-			return "deviceApproved";
+			model.addAttribute(APPROVED, false);
+			return tryToSetThemedApproveDevice(model, p, req, (DeviceCode) session.getAttribute(DEVICE_CODE));
 		}
 
 		// create an OAuth request for storage
 		OAuth2Request o2req = oAuth2RequestFactory.createOAuth2Request(authorizationRequest);
 		OAuth2Authentication o2Auth = new OAuth2Authentication(o2req, auth);
 		
 		DeviceCode approvedCode = deviceCodeService.approveDeviceCode(dc, o2Auth);
-		
-		// pre-process the scopes
+		Set<SystemScope> sortedScopes = getSortedScopes(dc);
+
+		model.put(SCOPES, sortedScopes);
+		model.put(APPROVED, true);
+
+		return DEVICE_APPROVED;
+	}
+
+	private String checkRequestIsValid(String clientId, ClientWithScopes clientWithScopes, String scope, ModelMap model) {
```

Comment:
Try to find a better name for the method. I would expect a check method to return true/false, based on the checks made on parameters. This method seems to do too much of mixed work - modifies `model` object, modifies `clientWithScopes` parameter, etc.